accel/tcg/log_instr.c

```diff
@@ -0,0 +1,902 @@
+ */
+
+#include <stdint.h>
+
+#include "qemu/osdep.h"
+#include "qemu/range.h"
+#include "qemu/log.h"
+#include "cpu.h"
+#include "exec/exec-all.h"
+#include "exec/log.h"
+#include "exec/helper-proto.h"
+#include "exec/log_instr.h"
+#include "exec/memop.h"
+#include "disas/disas.h"
+
+/*
+ * CHERI common instruction logging.
+ *
+ * This is the central implementation of the CPU_LOG_INSTR and CPU_LOG_USER_ONLY
+ * tracing.
+ * The same functions can be used by CHERI targets to append to the instruction
+ * log buffer. Once the instruction is fully processed, the target commits the
+ * log buffer and depending on the instruction operations and -dfilter options
+ * we either flush the buffer or drop it.
+ * A central design goal is to deterministically output the main side effect of
+ * the instruction (generally the destination register value), other side
+ * effects, such as multiple register updates, may be implemented by emitting
+ * multiple trace entries, but this is currently unimplemented.
+ * Another design goal is to allow tracing of arbitrary events via nops in
+ * both the text and binary format.
+ *
+ * TODO(am2419): how do we deal with orderding in case multiple registers are updated?
+ * This is critical to recognize which value goes in which register, and also how to
+ * tie multiple memory accesses to the respective value/register.
+ * We could add an explicit target-specific register ID handle in place of the
+ * register name. This could be used also to fetch the register name and would
+ * provide an identifier to external parsers.
+ * Memory updates are harder to deal with, at least in the current format, perhaps
+ * the semantic of the instruction is enough to recover the ordering from a trace.
+ */
+
+#ifdef CONFIG_TCG_LOG_INSTR
+
+#ifdef CONFIG_DEBUG_TCG
+#define log_assert(x) assert((x))
+#else
+#define log_assert(x)
+#endif
+
+/*
+ * -dfilter ranges in common logging implementation.
+ */
+extern GArray *debug_regions;
+
+struct log_reginfo {
+    uint16_t flags;
+#define LRI_CAP_REG    1
+#define LRI_HOLDS_CAP  2
+
+    const char *name;
+    union {
+        target_ulong gpr;
+#ifdef TARGET_CHERI
+        cap_register_t cap;
+#endif
+    };
+};
+typedef struct log_reginfo log_reginfo_t;
+
+#define reginfo_is_cap(ri) (ri->flags & LRI_CAP_REG)
+#define reginfo_has_cap(ri) (reginfo_is_cap(ri) && (ri->flags & LRI_HOLDS_CAP))
+
+struct log_meminfo {
+    uint8_t flags;
+#define LMI_LD 1
+#define LMI_ST 2
+#define LMI_CAP 4
+    MemOp op;
+    target_ulong addr;
+    union {
+        uint64_t value;
+#ifdef TARGET_CHERI
+        cap_register_t cap;
+#endif
+    };
+};
+typedef struct log_meminfo log_meminfo_t;
+
+struct trace_fmt_hooks {
+    void (*emit_header)(CPUArchState *env);
+    void (*emit_start)(CPUArchState *env, target_ulong pc);
+    void (*emit_stop)(CPUArchState *env, target_ulong pc);
+    void (*emit_entry)(CPUArchState *env, cpu_log_instr_info_t *log);
+};
+typedef struct trace_fmt_hooks trace_fmt_hooks_t;
```

Comment:
You could merge these typedefs with the struct delcaration? `typdef struct foo { ... } foo_t;`.